crates/genie-rec/src/element.rs

```diff
@@ -0,0 +1,45 @@
+use crate::reader::RecordingHeaderReader;
+use std::io::{Read, Write};
+
+pub trait OptionalReadableElement: Sized {
+    fn read_from<R: Read>(input: &mut RecordingHeaderReader<R>) -> crate::Result<Option<Self>>;
+}
+
+impl<T: OptionalReadableElement> ReadableElement<Option<T>> for T {
+    fn read_from<R: Read>(input: &mut RecordingHeaderReader<R>) -> crate::Result<Option<T>> {
+        OptionalReadableElement::read_from(input)
+    }
+}
+
+pub trait ReadableElement<T>: Sized {
+    fn read_from<R: Read>(input: &mut RecordingHeaderReader<R>) -> crate::Result<T>;
+}
+
+pub trait ReadableHeaderElement: Sized {
+    fn read_from<R: Read>(input: &mut RecordingHeaderReader<R>) -> crate::Result<Self>;
+}
+
+impl<T: ReadableHeaderElement> ReadableElement<T> for T {
+    fn read_from<R: Read>(input: &mut RecordingHeaderReader<R>) -> crate::Result<T> {
+        ReadableHeaderElement::read_from(input)
+    }
+}
+
+pub trait WritableElement<T> {
+    fn write_to<W: Write>(element: &T, output: &mut W) -> crate::Result<()>;
+}
+
+pub trait WritableHeaderElement {
+    fn write_to<W: Write>(&self, output: &mut W) -> crate::Result<()> {
+        // we need to use `output` otherwise we'll get warnings that it's not used
+        // prefixing it would make any traits auto completed also be prefixed with _ and that's annoying
+        let _ = output;
+        unimplemented!()
+    }
```

Comment:
I hope you guys don't mind me chiming in :slightly_smiling_face: 
What is the point in providing a default trait implementation that just panics? Wouldn't it be better to just not provide a default trait implementation and add the `unimplemented` in the specific implementations?
```suggestion
    fn write_to<W: Write>(&self, output: &mut W) -> crate::Result<()>;
```

Reply:
@the-eater What's your opinion on this?

Reply:
Probably, it's OK like that instead of copying and pasting this to several places (DRY). Also, in the future, we can implement it in the trait as a default and then apply changes to the other local implementations of `write_to`.